lib/cn/node_split.c

```diff
@@ -366,3 +370,85 @@ cn_tree_node_get_split_key(
 
     return err;
 }
+
+static void
+kvset_split_res_init(struct cn_compaction_work *w, struct kvset_split_res *result, uint ks_idx)
+{
+    memset(result, 0, sizeof(*result));
+
+    for (int i = LEFT; i <= RIGHT; i++) {
+        uint idx = (i == LEFT ? ks_idx : ks_idx + w->cw_kvset_cnt);
+
+        result->ks[i].blks = &w->cw_outv[idx];
+        blk_list_init(&result->ks[i].blks->kblks);
+        blk_list_init(&result->ks[i].blks->vblks);
+        result->ks[i].blks->bl_vused = 0;
+
+        result->ks[i].blks_commit = &w->cw_split.commit[idx];
+        blk_list_init(result->ks[i].blks_commit);
+
+        result->ks[i].vgmap = &w->cw_vgmap[idx];
+    }
+
+    result->blks_purge = &w->cw_split.purge[ks_idx];
+    blk_list_init(result->blks_purge);
+}
+
+static void
+kvset_split_res_free(struct kvset *ks, struct kvset_split_res *result)
+{
+    for (int i = LEFT; i <= RIGHT; i++) {
+        blk_list_free(&result->ks[i].blks->kblks);
+        blk_list_free(&result->ks[i].blks->vblks);
+        blk_list_free(result->ks[i].blks_commit);
+    }
+
+    blk_list_free(result->blks_purge);
+}
+
+merr_t
+cn_split(struct cn_compaction_work *w)
+{
+    struct kvset_list_entry *le;
+    struct key_obj split_kobj;
+    struct cndb *cndb = cn_tree_get_cndb(w->cw_tree);
+    merr_t err;
+    uint i;
+
+    err = cn_tree_node_get_split_key(w->cw_node, w->cw_split.key, HSE_KVS_KEY_LEN_MAX,
+                                     &w->cw_split.klen);
+    if (err)
+        return err;
+
+    key2kobj(&split_kobj, w->cw_split.key, w->cw_split.klen);
+
+    for (i = 0, le = list_first_entry_or_null(&w->cw_node->tn_kvset_list, typeof(*le), le_link);
```

Comment:
declare i here

Reply:
Would mess with le.

Reply:
Funny!  I've never been a fan of for loops that span multiple lines, but maybe that's just me...

You can't be here if (cw_kvset_cnt < 1), right?  In which case you can use le = list_first_entry().

Reply:
Yeah, you're right. spe_work() has n_kvsets == 0 check. Will change this to use list_first_entry.